ext/dom/xpath.c

```diff
@@ -438,47 +439,12 @@ PHP_METHOD(DOMXPath, registerPhpFunctionNS)
 /* {{{ */
 PHP_METHOD(DOMXPath, quote) {
 	const char *input;
-	size_t input_len;
+	uintptr_t input_len;
 	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &input, &input_len) == FAILURE) {
 		RETURN_THROWS();
 	}
-	if (memchr(input, '\'', input_len) == NULL) {
-		zend_string *const output = zend_string_safe_alloc(1, input_len, 2, false);
-		output->val[0] = '\'';
-		memcpy(output->val + 1, input, input_len);
-		output->val[input_len + 1] = '\'';
-		output->val[input_len + 2] = '\0';
-		RETURN_STR(output);
-	} else if (memchr(input, '"', input_len) == NULL) {
-		zend_string *const output = zend_string_safe_alloc(1, input_len, 2, false);
-		output->val[0] = '"';
-		memcpy(output->val + 1, input, input_len);
-		output->val[input_len + 1] = '"';
-		output->val[input_len + 2] = '\0';
-		RETURN_STR(output);
-	} else {
-		smart_str output = {0};
-		// need to use the concat() trick published by Robert Rossney at https://stackoverflow.com/a/1352556/1067003
-		smart_str_appendl(&output, "concat(", 7);
-		const char *ptr = input;
-		const char *const end = input + input_len;
-		while (ptr < end) {
-			const char *const single_quote_ptr = memchr(ptr, '\'', end - ptr);
-			const char *const double_quote_ptr = memchr(ptr, '"', end - ptr);
-			const size_t distance_to_single_quote = single_quote_ptr ? single_quote_ptr - ptr : end - ptr;
-			const size_t distance_to_double_quote = double_quote_ptr ? double_quote_ptr - ptr : end - ptr;
-			const size_t bytes_until_quote = MAX(distance_to_single_quote, distance_to_double_quote);
-			const char quote_method = (distance_to_single_quote > distance_to_double_quote) ? '\'' : '"';
-			smart_str_appendc(&output, quote_method);
-			smart_str_appendl(&output, ptr, bytes_until_quote);
-			smart_str_appendc(&output, quote_method);
-			ptr += bytes_until_quote;
-			smart_str_appendc(&output, ',');
-		}
-		ZEND_ASSERT(ptr == end);
-		output.s->val[output.s->len - 1] = ')';
-		RETURN_STR(smart_str_extract(&output));
-	}
+	const char *ouput = domxpath_quote_literal(input, &input_len);
+	RETURN_STRINGL(ouput, input_len);
```

Comment:
Typo: `ouput`. This construct will also require a full allocation + copy into a new `zend_string()`.